File: python/massmarket_hash_event/hamt.py
# SPDX-FileCopyrightText: 2025 Mass Labs
#
# SPDX-License-Identifier: MIT

import xxhash
from dataclasses import dataclass
from typing import TypeVar, Generic, Optional, Any, Callable
import cbor2

BITS_PER_STEP = 5
MAX_DEPTH = (64 + BITS_PER_STEP - 1) // BITS_PER_STEP

V = TypeVar("V")


@dataclass
class HashState:
    original_key: bytes
    hash: int
    consumed: int
    seed: int

    @classmethod
    def new(cls, key: bytes) -> "HashState":
        return cls(
            original_key=key, hash=hash_key_with_seed(key, 0), consumed=0, seed=0
        )

    def next(self) -> int:
        if self.consumed + BITS_PER_STEP > MAX_DEPTH * BITS_PER_STEP:
            self.seed += 1
            self.hash = hash_key_with_seed(self.original_key, self.seed)
            self.consumed = 0

        shift = self.consumed
        mask = (1 << BITS_PER_STEP) - 1
        chunk = (self.hash >> shift) & mask
        self.consumed += BITS_PER_STEP
        return chunk


@dataclass
class Entry(Generic[V]):
    key: Optional[bytes]
    value: Optional[V]
    node: Optional["Node[V]"]


@dataclass
class Node(Generic[V]):
    bitmap: int = 0
    entries: list[Entry[V]] = None
    _hash: Optional[bytes] = None

    def __post_init__(self):
        if self.entries is None:
            self.entries = []

    def insert(self, key: bytes, value: V, hs: HashState) -> bool:
        if hs.consumed >= MAX_DEPTH * BITS_PER_STEP:
            return self.insert_fallback(key, value)

        idx = hs.next()
        if idx >= 32:
            raise ValueError(f"idx out of range: {idx}")

        pos = count_ones(self.bitmap, idx)

        if self.bitmap & (1 << idx) == 0:
            self.bitmap |= 1 << idx
            self.entries.insert(pos, Entry(key=key, value=value, node=None))
            self._hash = None
            return True

        entry = self.entries[pos]
        if entry.node is None:
            if entry.key == key:
                entry.value = value
                self._hash = None
                return False

            branch = Node()
            old_hs = HashState(
                original_key=entry.key,
                hash=hash_key(entry.key),
                consumed=hs.consumed,
                seed=hs.seed,
            )
            branch.insert(entry.key, entry.value, old_hs)

            new_hs = HashState(
                original_key=key, hash=hash_key(key), consumed=hs.consumed, seed=hs.seed
            )
            branch.insert(key, value, new_hs)

            self.entries[pos] = Entry(key=None, value=None, node=branch)
            self._hash = None
            return True

        inserted = entry.node.insert(key, value, hs)
        if inserted:
            self._hash = None
        return inserted

    def find(self, key: bytes) -> tuple[Optional[V], bool]:
        hs = HashState.new(key)
        current_node = self

        while True:
            if current_node is None:
                return None, False

            if hs.consumed >= MAX_DEPTH * BITS_PER_STEP:
                return current_node.find_fallback(key)

            idx = hs.next()
            if current_node.bitmap & (1 << idx) == 0:
                return None, False

            pos = count_ones(current_node.bitmap, idx)
            if pos >= len(current_node.entries):
                return None, False

            entry = current_node.entries[pos]
            if entry.node is None:
                if entry.key == key:
                    return entry.value, True
                return None, False

            current_node = entry.node

    def delete(self, key: bytes, hs: HashState) -> bool:
        if hs.consumed >= MAX_DEPTH * BITS_PER_STEP:
            return self.delete_fallback(key)

        idx = hs.next()
        if self.bitmap & (1 << idx) == 0:
            return False

        pos = count_ones(self.bitmap, idx)
        if pos >= len(self.entries):
            raise ValueError(f"pos for idx {idx} out of range: {pos}")

        entry = self.entries[pos]
        if entry.node is None:
            if entry.key != key:
                return False

            self._hash = None
            self.bitmap &= ~(1 << idx)
            self.entries.pop(pos)
            return True

        deleted = entry.node.delete(key, hs)
        if not deleted:
            return False

        if len(entry.node.entries) == 0:
            self.bitmap &= ~(1 << idx)
            self.entries.pop(pos)
        elif len(entry.node.entries) == 1 and entry.node.entries[0].node is None:
            child_entry = entry.node.entries[0]
            self.entries[pos] = Entry(
                key=child_entry.key, value=child_entry.value, node=None
            )

        self._hash = None
        return True

    def insert_fallback(self, key: bytes, value: V) -> bool:
        for i, e in enumerate(self.entries):
            if e.key == key:
                self.entries[i].value = value
                return False
        self.entries.append(Entry(key=key, value=value, node=None))
        self._hash = None
        return True

    def find_fallback(self, key: bytes) -> tuple[Optional[V], bool]:
        for e in self.entries:
            if e.node is None and e.key == key:
                return e.value, True
            elif e.node is not None:
                value, found = e.node.find_fallback(key)
                if found:
                    return value, True
        return None, False

    def delete_fallback(self, key: bytes) -> bool:
        for i, e in enumerate(self.entries):
            if e.node is None and e.key == key:
                self.entries.pop(i)
                self._hash = None
                return True
        return False

    def all(self, fn: Callable[[bytes, V], bool]) -> bool:
        if self is None:
            return True
        for e in self.entries:
            if e.node is None:
                if not fn(e.key, e.value):
                    return False
            else:
                if not e.node.all(fn):
                    return False
        return True

    def hash(self) -> bytes:
        if self._hash is not None:
            return self._hash

        h = xxhash.xxh64()
        for e in self.entries:
            if e.node is None:
                h.update(e.key)
                h.update(cbor2.dumps(e.value))
            else:
                h.update(e.node.hash())

        self._hash = h.digest()
        return self._hash


def hash_key(key: bytes) -> int:
    return xxhash.xxh64(key).intdigest()


def hash_key_with_seed(key: bytes, seed: int) -> int:
    return xxhash.xxh64(key, seed=seed).intdigest()


def count_ones(n: int, below: int) -> int:
    mask = (1 << below) - 1
    return bin(n & mask).count("1")


def deep_copy_node(node: Node[V]) -> Node[V]:
    if node is None:
        return None

    new_node = Node(bitmap=node.bitmap)
    new_node.entries = []

    for entry in node.entries:
        new_entry = Entry(
            key=entry.key,
            value=entry.value,
            node=deep_copy_node(entry.node) if entry.node else None,
        )
        new_node.entries.append(new_entry)

    return new_node


@dataclass
class Trie(Generic[V]):
    root: Node[V]
    size: int = 0

    @classmethod
    def new(cls) -> "Trie[V]":
        return cls(root=Node())

    def insert(self, key: bytes, value: V) -> None:
        if self.root is None:
            self.root = Node()

        inserted = self.root.insert(key, value, HashState.new(key))
        if inserted:
            self.size += 1

    def get(self, key: bytes) -> tuple[Optional[V], bool]:
        if self.root is None:
            return None, False
        return self.root.find(key)

    def delete(self, key: bytes) -> None:
        if self.root is None:
            return
        deleted = self.root.delete(key, HashState.new(key))
        if deleted:
            self.size -= 1

    def all(self, fn: Callable[[bytes, V], bool]) -> None:
        if self.root is None:
            return
        self.root.all(fn)

    def hash(self) -> bytes:
        if self.root is None:
            return xxhash.xxh64().digest()
        return self.root.hash()

    def copy(self) -> "Trie[V]":
        """Create a deep copy of the trie"""
        new_trie = Trie.new()
        new_trie.root = deep_copy_node(self.root)
        new_trie.size = self.size
        return new_trie

File: go/cbor/hamt.go
// SPDX-FileCopyrightText: 2024 - 2025 Mass Labs
//
// SPDX-License-Identifier: MIT

package schema

import (
	"bytes"
	"fmt"
	"io"
	"math/bits"
	"slices"

	"github.com/cespare/xxhash/v2"
)

const (
	bitsPerStep = 5
	maxDepth    = (64 + bitsPerStep - 1) / bitsPerStep
)

var (
	newHAMTHashFn = xxhash.New
	nilHAMTHash   []byte
)

func init() {
	nilHAMTHash = newHAMTHashFn().Sum(nil)
}

type hashState struct {
	originalKey []byte
	hash        uint64
	consumed    uint8
	seed        uint64
}

func newHashState(key []byte) *hashState {
	return &hashState{
		originalKey: key,
		hash:        hashKeyWithSeed(key, 0),
		consumed:    0,
		seed:        0,
	}
}

// next returns the next bitsPerStep bits from the hash state
func (hs *hashState) next() uint32 {
	if hs.consumed+bitsPerStep > maxDepth*bitsPerStep {
		hs.seed++
		hs.hash = hashKeyWithSeed(hs.originalKey, hs.seed)
		hs.consumed = 0
	}

	shift := hs.consumed
	mask := uint64((1 << bitsPerStep) - 1)
	chunk := (hs.hash >> shift) & mask
	hs.consumed += bitsPerStep
	return uint32(chunk)
}

var (
	// variables for test stubbing of hash collisions
	hashKeyWithSeed = hashKeyWithSeedFunc
	hashKey         = hashKeyFunc
)

// hashKey is used to hash keys
func hashKeyFunc(key []byte) uint64 {
	h := newHAMTHashFn()
	h.Write(key)
	return h.Sum64()
}

// hashKeyWithSeed is used when hashState consumed all bits
func hashKeyWithSeedFunc(key []byte, seed uint64) uint64 {
	// TODO: this uses the seed variation not newHashFn
	h := xxhash.NewWithSeed(seed)
	h.Write(key)
	return h.Sum64()
}

// Trie represents a persistent HAMT
type Trie[V any] struct {
	root *Node[V]
	size int
}

// Node is a node in the HAMT
type Node[V any] struct {
	_       struct{} `cbor:",toarray"`
	Bitmap  uint32
	Entries []Entry[V]

	// not serialized
	hash []byte
}

// Entry is either a leaf node (Node == nil) or a branch node (Node != nil)
type Entry[V any] struct {
	_     struct{} `cbor:",toarray"`
	Key   []byte
	Value V
	Node  *Node[V]
}

// NewTrie creates a new empty HAMT
func NewTrie[V any]() *Trie[V] {
	return &Trie[V]{
		root: &Node[V]{},
		size: 0,
	}
}

// MarshalCBOR marshals the HAMT into a CBOR encoded byte slice
func (t *Trie[V]) MarshalCBOR() ([]byte, error) {
	// use the schema package to inherit the canonical encoding options
	return Marshal(t.root)
}

// UnmarshalCBOR unmarshals a CBOR encoded byte slice into a HAMT
func (t *Trie[V]) UnmarshalCBOR(data []byte) error {
	if err := Unmarshal(data, &t.root); err != nil {
		return err
	}

	// Recalculate size
	size := 0
	var countEntries func(*Node[V])
	countEntries = func(n *Node[V]) {
		if n == nil {
			return
		}
		for _, e := range n.Entries {
			if e.Node == nil {
				size++ // Count direct entries
				continue
			}
			countEntries(e.Node)
		}
	}
	countEntries(t.root)
	t.size = size

	return nil
}

// Insert inserts a key-value pair into the HAMT
func (t *Trie[V]) Insert(key []byte, value V) error {
	if t.root == nil {
		t.root = &Node[V]{}
	}

	inserted, err := t.root.insert(key, value, newHashState(key))
	if err != nil {
		return err
	}

	if inserted {
		t.size++
	}
	return nil
}

func (n *Node[V]) insert(key []byte, value V, hs *hashState) (bool, error) {
	var nilValue V
	if hs.consumed >= maxDepth*bitsPerStep {
		return n.insertFallback(key, value)
	}

	// figure out which position in the entries array to insert the new entry
	idx := hs.next()
	if idx >= 32 {
		return false, fmt.Errorf("idx out of range: %d", idx)
	}
	pos := bits.OnesCount32(n.Bitmap & ((1 << idx) - 1))

	// if the position is not occupied, insert the new entry
	if n.Bitmap&(1<<idx) == 0 {
		n.Bitmap |= (1 << idx) // flip bit to 1
		n.Entries = append(n.Entries, Entry[V]{})
		// shift entries to the right to make space for the new entry
		if pos < len(n.Entries)-1 {
			copy(n.Entries[pos+1:], n.Entries[pos:])
		}
		n.Entries[pos] = Entry[V]{Key: key, Value: value}
		// invalidate the cached hash since we are mutating the node
		n.hash = nil
		return true, nil
	}

	// if the position is occupied, check if it's a branch node
	entry := &n.Entries[pos]
	if entry.Node == nil {
		// if it's not a branch node, check if the key is already in the entry
		if bytes.Equal(entry.Key, key) {
			var currentHash []byte
			if n.hash != nil {
				currentHash = n.hash
			} else {
				var h1 = newHAMTHashFn()
				err := n.encodeValue(entry.Value, h1)
				if err != nil {
					return false, err
				}
				currentHash = h1.Sum(nil)
			}
			var h2 = newHAMTHashFn()
			err := n.encodeValue(value, h2)
			if err != nil {
				return false, err
			}
			newHash := h2.Sum(nil)
			if bytes.Equal(currentHash, newHash) {
				return false, nil
			}
			// update the value directly
			entry.Value = value
			// invalidate the cached hash since we are mutating the node
			n.hash = nil
			return false, nil
		}

		// not a branch node yet, so create a new branch node
		branch := &Node[V]{}

		// create a new hash state for the existing key
		oldHS := &hashState{
			originalKey: entry.Key,
			hash:        hashKey(entry.Key),
			consumed:    hs.consumed,
			seed:        hs.seed,
		}
		_, err := branch.insert(entry.Key, entry.Value, oldHS)
		if err != nil {
			return false, err
		}

		// create a new hash state for the new key
		newHS := &hashState{
			originalKey: key,
			hash:        hashKey(key),
			consumed:    hs.consumed,
			seed:        hs.seed,
		}
		_, err = branch.insert(key, value, newHS)
		if err != nil {
			return false, err
		}

		// update the current node
		entry.Node = branch
		entry.Key = nil
		entry.Value = nilValue

		// invalidate the cached hash since we are mutating the node
		n.hash = nil
		return true, nil
	}

	// it is a branch node, so recursively insert the new entry
	inserted, err := entry.Node.insert(key, value, hs)
	if inserted {
		// invalidate the cached hash since we are mutating the node
		n.hash = nil
	}
	return inserted, err
}

// insertFallback is used when the hash state has consumed all bits
func (n *Node[V]) insertFallback(key []byte, value V) (bool, error) {
	for i, e := range n.Entries {
		if bytes.Equal(e.Key, key) {
			n.Entries[i].Value = value
			return false, nil
		}
	}

	// if the key is not found, append it to the entries
	n.Entries = append(n.Entries, Entry[V]{Key: key, Value: value})
	// invalidate the cached hash since we are mutating the node
	n.hash = nil
	return true, nil
}

// Get retrieves the value associated with a key from the HAMT
func (t *Trie[V]) Get(key []byte) (V, bool) {
	if t.root == nil {
		var nilValue V
		return nilValue, false
	}
	return t.root.Find(key)
}

// Find finds a given entry in the trie
func (n *Node[V]) Find(key []byte) (V, bool) {
	var nilValue V
	hs := newHashState(key)
	currentNode := n

	// iterate instead of recursion calls to avoid stack overflow
	for {
		if currentNode == nil {
			return nilValue, false
		}

		// are we at max depth?
		if hs.consumed >= maxDepth*bitsPerStep {
			return currentNode.findFallback(key)
		}

		// figure out which position in the entries array to look for the key
		idx := hs.next()
		if currentNode.Bitmap&(1<<idx) == 0 {
			return nilValue, false
		}

		// figure out which position in the entries array the key is at
		pos := bits.OnesCount32(currentNode.Bitmap & ((1 << idx) - 1))
		if pos >= len(currentNode.Entries) { // invalid position
			return nilValue, false
		}
		entry := &currentNode.Entries[pos]

		// if the entry is not a branch node, check if the key is in the entry
		if entry.Node == nil {
			if bytes.Equal(entry.Key, key) {
				return entry.Value, true
			}
			return nilValue, false
		}

		// if the entry is a branch node, recursively search the branch
		currentNode = entry.Node
	}
}

// findFallback is used when the hash state has consumed all bits
func (n *Node[V]) findFallback(key []byte) (V, bool) {
	for _, e := range n.Entries {
		if e.Node == nil && bytes.Equal(e.Key, key) {
			return e.Value, true
		} else if e.Node != nil {
			if val, found := e.Node.findFallback(key); found {
				return val, true
			}
		}
	}
	var nilValue V
	return nilValue, false
}

// Delete deletes a key from the HAMT
func (t *Trie[V]) Delete(key []byte) error {
	if t.root == nil {
		return nil
	}
	deleted, err := t.root.delete(key, newHashState(key))
	if err != nil {
		return err
	}
	if deleted {
		t.size--
	}
	return nil
}

func (n *Node[V]) delete(key []byte, hs *hashState) (bool, error) {
	// are we at max depth?
	if hs.consumed >= maxDepth*bitsPerStep {
		return n.deleteFallback(key)
	}

	// figure out which position in the entries array the key is at
	idx := hs.next()
	if n.Bitmap&(1<<idx) == 0 {
		return false, nil
	}

	pos := bits.OnesCount32(n.Bitmap & ((1 << idx) - 1))
	if pos >= len(n.Entries) {
		return false, fmt.Errorf("pos for idx %d out of range: %d", idx, pos)
	}

	entry := &n.Entries[pos]

	// if the entry is not a branch node, check if the key is in the entry
	if entry.Node == nil {
		if !bytes.Equal(entry.Key, key) {
			return false, nil
		}

		// invalidate hash since we are mutating the node
		n.hash = nil

		// Remove entry in place
		n.Bitmap &= ^(1 << idx)
		n.Entries = slices.Delete(n.Entries, pos, pos+1)
		return true, nil
	}

	// Handle branch node
	deleted, err := entry.Node.delete(key, hs)
	if err != nil {
		return false, err
	}
	if !deleted {
		return false, nil
	}

	// If child is empty, remove it
	if len(entry.Node.Entries) == 0 {
		n.Bitmap &= ^(1 << idx)
		n.Entries = slices.Delete(n.Entries, pos, pos+1)
	} else if len(entry.Node.Entries) == 1 && entry.Node.Entries[0].Node == nil {
		// If child has only one entry, collapse it
		childEntry := entry.Node.Entries[0]
		entry.Key = childEntry.Key
		entry.Value = childEntry.Value
		entry.Node = nil
	}

	// invalidate hash since we are mutating the node
	n.hash = nil

	return true, nil
}

func (n *Node[V]) deleteFallback(key []byte) (bool, error) {
	for i, e := range n.Entries {
		if e.Node == nil && bytes.Equal(e.Key, key) {
			n.Entries = slices.Delete(n.Entries, i, i+1)
			// invalidate the cached hash since we are mutating the node
			n.hash = nil
			return true, nil
		}
	}
	return false, nil
}

// Size returns the number of entries in the HAMT
func (t *Trie[V]) Size() int {
	return t.size
}

// Hash hashes the whole HAMT
func (t *Trie[V]) Hash() ([]byte, error) {
	if t.root == nil {
		return nilHAMTHash, nil
	}
	return t.root.Hash()
}

func (n *Node[V]) Hash() ([]byte, error) {
	h := newHAMTHashFn()
	if n == nil {
		return h.Sum(nil), nil
	}

	if n.hash != nil {
		return n.hash, nil
	}

	// Hash entries in order
	for _, e := range n.Entries {
		if e.Node == nil {
			h.Write(e.Key)
			err := n.encodeValue(e.Value, h)
			if err != nil {
				return nil, err
			}
		} else {
			hash, err := e.Node.Hash()
			if err != nil {
				return nil, err
			}
			h.Write(hash)
		}
	}

	n.hash = h.Sum(nil)
	return n.hash, nil
}

func (n *Node[V]) encodeValue(v V, w io.Writer) error {
	enc := DefaultEncoder(w)
	return enc.Encode(v)
}

// All iterates over all key-value pairs in the trie
// The iteration stops if fn returns false.
func (t *Trie[V]) All(fn func([]byte, V) bool) {
	if t.root == nil {
		return
	}
	t.root.all(fn)
}

func (n *Node[V]) all(fn func([]byte, V) bool) bool {
	if n == nil {
		return true
	}
	for _, e := range n.Entries {
		if e.Node == nil {
			if !fn(e.Key, e.Value) {
				return false
			}
		} else {
			if !e.Node.all(fn) {
				return false
			}
		}
	}
	return true
}
